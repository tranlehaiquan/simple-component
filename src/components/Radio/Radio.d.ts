import * as React from 'react';

export interface RadioProps {
  children?: React.ReactNode,
  checked?: boolean,
  onChange? (value: number | string | boolean, event: React.SyntheticEvent): void,
  value: number | string | boolean,
  name: string,
  disabled?: boolean,
  required?: boolean
}

declare const Radio: React.ComponentType<RadioProps>;

export default Radio;
