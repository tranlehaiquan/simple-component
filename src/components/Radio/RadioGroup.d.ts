import * as React from 'react';

export interface RadioGroupProps {
  children: React.ReactNode,
  value: number | string | boolean,
  name: string,
  onChange: (value: number | string | boolean, event: React.SyntheticEvent) => void,
  required: boolean,
  disabled: boolean
}

declare const RadioGroup: React.ComponentType<RadioGroupProps>;

export default RadioGroup;
