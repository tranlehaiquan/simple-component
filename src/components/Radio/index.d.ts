export { default as RadioGroup } from './RadioGroup';

export { default } from './Radio';
export * from './Radio';
