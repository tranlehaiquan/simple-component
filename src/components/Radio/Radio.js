import React, { Component } from 'react';
import propTypes from 'prop-types';
import classnames from 'classnames';

class Radio extends Component {
	handleOnChange = () => {
		const { onChange, value } = this.props;

		onChange(value);
	}
	
	handleKeyPress = (e) => {
		const keyCodeOfSpace = 32;

		if (e.keyCode !== keyCodeOfSpace) return;

		const { onChange, value } = this.props;
		onChange(value);
	}

	render() {
		const { children, checked, value, name, disabled, ...restProps } = this.props;
		const classnameRadio = classnames('sp-radio__item', {
			['sp-radio__item--checked']: checked,
			['sp-radio__item--disabled']: disabled 
		});

		return (
			<label className={classnameRadio}>
				<input 
					type="radio"
					className="sp-radio__input" 
					name={name} 
					value={value} 
					tabIndex="-1"
					checked={checked}
					onChange={this.handleOnChange}
					disabled={disabled}
					{...restProps}
				/>
				<span 
					className="sp-radio__icon" 
					tabIndex={disabled ? '-1' : '0'}
					onKeyDown={this.handleKeyPress} 
				/>
				<span className="sp-radio__text">{ children }</span>
			</label>
		);
	}
}

Radio.propTypes = {
	children: propTypes.node,
	checked: propTypes.bool,
	onChange: propTypes.func,
	value: propTypes.oneOfType([
		propTypes.number, 
		propTypes.string, 
		propTypes.bool
	]),
	name: propTypes.string,
	disabled: propTypes.bool
};

Radio.defaultProps = {
	checked: false,
	disabled: false
};

export default Radio;
