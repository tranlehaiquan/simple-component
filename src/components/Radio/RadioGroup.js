import React, {
  memo
} from 'react';
import propTypes from 'prop-types';

function RadioGroup(props) {
  function handleOnClick(value, e) {
    props.onChange(value, e);
  }

  function renderChildrent() {
    const {
      children,
      value,
      name
    } = props;

    const radios = React.Children.map(children, (radio) => {
      const checked = value == radio.props.value;

      return React.cloneElement(
        radio, {
          checked,
          onChange: handleOnClick,
          name
        }
      );
    });

    return radios;
  }

  return ( 
    <div className = "sp-radio__group" > 
      {
        renderChildrent()
      } 
    </div>
  );
}

RadioGroup.propTypes = {
  children: propTypes.node,
  value: propTypes.oneOfType([
    propTypes.number,
    propTypes.string,
    propTypes.bool
  ]),
  name: propTypes.string,
  onChange: propTypes.func,
  required: propTypes.bool,
  disabled: propTypes.bool
};

RadioGroup.defaultProps = {
  name: '',
  value: '',
  required: false,
  disabled: false
};

export default memo(RadioGroup);
