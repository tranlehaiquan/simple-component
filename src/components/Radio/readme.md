# todo

- Add behave focus -> space keypress -> change value
