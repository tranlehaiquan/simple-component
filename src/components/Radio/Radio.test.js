import React from 'react';
import renderer from 'react-test-renderer';
import Radio from './Radio';
import RadioGroup from './RadioGroup';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';

describe('make snapshot', () => {
  test('Render Radios with, radio group', () => {
    const component = renderer.create(
      <RadioGroup 
        name="hello"
        value={1}
        onChange={() => {}}
      >
        <Radio value="1">Option 1</Radio>
        <Radio value="2" disabled>Option 2</Radio>
        <Radio value="3">Option 3</Radio>
      </RadioGroup>
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  test('Render single Radio', () => {
    const component = renderer.create(
      <Radio value="1">Option 1</Radio>
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Test Radio with props', () => {
  test('Normal radio', () => {
    const onRadioClick = sinon.spy();
    const component = shallow(<Radio value="1" onChange={onRadioClick} >Value 1</Radio>);
    const input = component.find('input');
  
    expect(component.find('.sp-radio__icon')).toBeDefined();
    expect(component.text()).toBe('Value 1');
    expect(input.prop('disabled')).toBeFalsy();
  
    input.simulate('change');
    expect(onRadioClick).toHaveProperty('callCount', 1);
    expect(onRadioClick.args[0][0]).toBe('1');
  });

  test('Radio disabled', () => {
    const onRadioClick = sinon.spy();
    const component = shallow(<Radio value="1" disabled onChange={onRadioClick}>Value 1</Radio>);
    const input = component.find('input');
    const icon = component.find('.sp-radio__icon');
  
    expect(input.prop('disabled')).toBeTruthy();
    expect(input.prop('value')).toBe('1');
    expect(icon.prop('tabIndex')).toBe('-1');
    
    input.simulate('click');
    expect(onRadioClick).toHaveProperty('callCount', 0);
  });

  test('Radio group with event onchange', () => {
    const onRadioChange = sinon.spy();
    const component = mount(
      <RadioGroup 
        name="hello"
        value={1}
        onChange={onRadioChange}
      >
        <Radio value="1">Option 1</Radio>
        <Radio value="2" disabled>Option 2</Radio>
        <Radio value="3">Option 3</Radio>
      </RadioGroup>
    );

    component.find('Radio').first().find('input').simulate('change');

    expect(onRadioChange).toHaveProperty('callCount', 1);
    expect(onRadioChange.args[0][0]).toBe('1');

    component.setProps({ value: 2 });
    expect(component.find('Radio').at(1).prop('checked')).toBeTruthy();
  });
});




