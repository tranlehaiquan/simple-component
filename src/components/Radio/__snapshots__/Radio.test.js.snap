// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Radios with, radio group 1`] = `
<div
  className="sp-radio__group"
>
  <label
    className="sp-radio__item sp-radio__item--checked"
  >
    <input
      checked={true}
      className="sp-radio__input"
      disabled={false}
      name="hello"
      onChange={[Function]}
      tabIndex="-1"
      type="radio"
      value="1"
    />
    <span
      className="sp-radio__icon"
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <span
      className="sp-radio__text"
    >
      Option 1
    </span>
  </label>
  <label
    className="sp-radio__item sp-radio__item--disabled"
  >
    <input
      checked={false}
      className="sp-radio__input"
      disabled={true}
      name="hello"
      onChange={[Function]}
      tabIndex="-1"
      type="radio"
      value="2"
    />
    <span
      className="sp-radio__icon"
      onKeyDown={[Function]}
      tabIndex="-1"
    />
    <span
      className="sp-radio__text"
    >
      Option 2
    </span>
  </label>
  <label
    className="sp-radio__item"
  >
    <input
      checked={false}
      className="sp-radio__input"
      disabled={false}
      name="hello"
      onChange={[Function]}
      tabIndex="-1"
      type="radio"
      value="3"
    />
    <span
      className="sp-radio__icon"
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <span
      className="sp-radio__text"
    >
      Option 3
    </span>
  </label>
</div>
`;

exports[`Render single Radio 1`] = `
<label
  className="sp-radio__item"
>
  <input
    checked={false}
    className="sp-radio__input"
    disabled={false}
    onChange={[Function]}
    tabIndex="-1"
    type="radio"
    value="1"
  />
  <span
    className="sp-radio__icon"
    onKeyDown={[Function]}
    tabIndex="0"
  />
  <span
    className="sp-radio__text"
  >
    Option 1
  </span>
</label>
`;
